domain/topn_slow_query.go

```diff
@@ -59,28 +62,87 @@ func (h *slowQueryHeap) Refresh(now time.Time, recent time.Duration) {
 	heap.Init(h)
 }
 
+func (h *slowQueryHeap) Query(count int) []*SlowQueryInfo {
+	save := make([]*SlowQueryInfo, len(h.data))
+	copy(save, h.data)
+
+	sort.Sort(h)
+
+	ret := make([]*SlowQueryInfo, 0, count)
+	for i := len(h.data) - 1; i >= 0 && len(ret) <= count; i-- {
+		ret = append(ret, h.data[i])
+	}
+
+	// Sort breaks the data, recover it here.
+	h.data = save
+	return ret
+}
+
+type slowQueryQueue struct {
+	data []*SlowQueryInfo
+	head int
+	tail int
+}
+
+func (q *slowQueryQueue) Enqueue(info *SlowQueryInfo) {
+	q.data[q.tail] = info
+	q.tail = (q.tail + 1) % len(q.data)
+	if q.tail == q.head {
+		q.head = (q.head + 1) % len(q.data)
+	}
+}
+
+func (q *slowQueryQueue) Query(count int) []*SlowQueryInfo {
+	// Queue is empty.
+	if q.tail == q.head {
+		return nil
+	}
+
+	if count > len(q.data) {
+		count = len(q.data)
+	}
+	ret := make([]*SlowQueryInfo, 0, count)
+	tail := (q.tail - 1 + len(q.data)) % len(q.data)
```

Comment:
This is simpler
```go
if q.head > q.tail {
  ret = append(ret, q.data[head:]...)
  ret = append(ret, q.data[:head]...)
} else {
  ret = append(ret, q.data[head:tail+1]...)
}
```

Reply:
It's not simpler, it brings more mental burden.